packages/signing-utils/src/signing-clients/index.ts

```diff
@@ -0,0 +1,83 @@
+import type { ConnectConfig } from 'ssh2';
+
+import * as path from 'path';
+import { SSHClient } from '../ssh-client';
+import { LocalSigningClient } from './local-signing-client';
+import { RemoteSigningClient } from './remote-signing-client';
+
+export { LocalSigningClient } from './local-signing-client';
+export { RemoteSigningClient } from './remote-signing-client';
+
+export type SigningMethod = 'gpg' | 'jsign';
+
+export type SigningClientOptions = {
+  workingDirectory: string;
+  signingScript: string;
+  signingMethod: SigningMethod;
+};
+
+/** Options for signing a file remotely over an SSH connection. */
+export type RemoteSigningOptions = {
+  /** Username for authentication. */
+  username?: string;
+  /** Password for password-based user authentication. */
+  password?: string;
+  /** Port number of the ssh server. */
+  port?: number;
+  /** Buffer or string that contains a private key for either key-based or hostbased user authentication (OpenSSH format). */
+  privateKey?: Buffer | string;
+  /** The method to sign with.  Use gpg on linux and jsign on windows. */
+  signingMethod: SigningMethod;
+  client: 'remote';
+};
+
+/** Options for signing a file locally. */
+export type LocalSigningOptions = {
+  /** The method to sign with.  Use gpg on linux and jsign on windows. */
+  signingMethod: SigningMethod;
+
+  /** Full path to the directory in which to produce the signed file. */
+  directory?: string;
+
+  client: 'local';
+};
+
+export type ClientOptions = RemoteSigningOptions | LocalSigningOptions;
+
+export interface SigningClient {
+  sign(file: string): Promise<void>;
+}
+
+export async function getSigningClient(
+  options: ClientOptions
+): Promise<SigningClient> {
+  async function getSshClient(sshOptions: ConnectConfig) {
+    const sshClient = new SSHClient(sshOptions);
+    await sshClient.connect();
+    return sshClient;
+  }
+
+  const signingScript = path.join(__dirname, '../..', 'src', './garasign.sh');
+
+  if (options.client === 'remote') {
+    const sshClient = await getSshClient(options);
+    // Currently only linux remote is supported to sign the artifacts
+    return new RemoteSigningClient(sshClient, {
+      workingDirectory: '~/garasign',
```

Comment:
I know this path value is here from my PR, but after debugging for some time, I realized this path is not expanded by default when copying files to the remote server. The sftp client interprets it exactly as this and throws error `No such file`. We need to expand it here to `/home/ubuntu/garasign` or some other absolute path.

Reply:
Could we use [realpath](https://github.com/mscdex/ssh2/blob/master/SFTP.md#:~:text=Error%20%3Eerr.-,realpath,-(%3C%20string%20%3Epath) to expand the working directory before we start using the RemoteClient?

Reply:
Seems like maybe an integration test would be valuable here.  What if we use ssh2 to spin up a testing ssh server to e2e test the remote client?  The only mechanism not tested in that case is the actual signing with Garasign.